uniffi_bindgen/src/crate_mode.rs

```diff
@@ -0,0 +1,242 @@
+#[derive(Debug)]
+pub struct Source {
+    pub package: Package,
+    pub crate_name: String,
+    pub ci: ComponentInterface,
+    pub config: Config,
+}
+
+fn calc_sources(
+    cargo_metadata: &cargo_metadata::Metadata,
+    library_path: &Utf8Path,
+) -> Result<Vec<Source>> {
+    let cdylib_name = library_path
+        .file_name()
+        .expect("Unexpected library path: {library_path}")
+        .strip_prefix(DLL_PREFIX)
+        .expect("Unexpected library path: {library_path}")
+        .strip_suffix(DLL_SUFFIX)
+        .expect("Unexpected library path: {library_path}");
+    group_metadata(macro_metadata::extract_from_library(library_path)?)?
+        .into_iter()
+        .map(|group| {
+            let package = find_package_by_crate_name(cargo_metadata, &group.namespace.crate_name)?;
+            let crate_root = package
+                .manifest_path
+                .parent()
+                .context("manifest path has no parent")?;
+            let mut ci =
+                load_component_interface(&group.namespace.crate_name, crate_root, &group.items)?;
+            let crate_name = group.namespace.crate_name.clone();
+            macro_metadata::add_group_to_ci(&mut ci, group)?;
+            let mut config = Config::load_initial(crate_root, None)?;
+            config.update_from_cdylib_name(cdylib_name);
+            config.update_from_ci(&ci);
+            Ok(Source {
+                config,
+                crate_name,
+                ci,
+                package,
+            })
+        })
+        .collect()
+}
+
+fn find_cdylib_path(main_crate_name: &str, package: &Package) -> Result<Utf8PathBuf> {
+    let cdylib_targets: Vec<&Target> = package
+        .targets
+        .iter()
+        .filter(|t| t.crate_types.iter().any(|t| t == "cdylib"))
+        .collect();
+    let target = match cdylib_targets.len() {
+        1 => cdylib_targets[0],
+        n => bail!("Found {n} cdylib targets for {}", package.name),
+    };
+
+    let messages = build_messages(main_crate_name)?;
+    let artifacts = messages.iter().filter_map(|message| match message {
+        Message::CompilerArtifact(artifact) => {
+            if artifact.target == *target {
+                Some(artifact.clone())
+            } else {
+                None
+            }
+        }
+        _ => None,
+    });
+    let cdylib_files: Vec<Utf8PathBuf> = artifacts
+        .into_iter()
+        .flat_map(|artifact| {
+            artifact
+                .filenames
+                .into_iter()
+                .filter(|nm| matches!(nm.extension(), Some(DLL_EXTENSION)))
+                .collect::<Vec<Utf8PathBuf>>()
+        })
+        .collect();
+
+    match cdylib_files.len() {
+        1 => Ok(cdylib_files[0].to_owned()),
+        n => bail!("Found {n} cdylib files for {}", package.name),
+    }
+}
+
+fn build_messages(main_crate_name: &str) -> Result<Vec<Message>> {
+    let mut child = Command::new(env!("CARGO"))
```

Comment:
In slack you mentioned you run `cargo build` to find the cdylib - isn't it possible that some people will have run their "real" (or "desired"?) `cargo build` with options that will make this be incorrect - eg, things like cross-compiling? I guess I'm wondering if it might be more robust to make the consumer specify this?

Reply:
I was thinking about this and I think it requires a different approach.  Maybe instead of the user specifying the crate and UniFFI building it, the user would need to specify the final library file.  It's a bit more work for them, but they're going to have to copy that library sooner or later, so I don't think it's that big of a deal.

Alternatively, we could make this command a wrapper around `cargo build` and let the user specify all the args they want to pass to it.  I'm not sure exactly how to do this though.